src/providers/json-rpc-provider.ts

```diff
@@ -152,32 +162,52 @@ export class JsonRpcProvider extends Provider {
-                    throw new TypedError(response.error.data.error_message, response.error.data.error_type);
-                } else {
-                    throw parseRpcError(response.error.data);
+        const result = await exponentialBackoff(REQUEST_RETRY_WAIT, REQUEST_RETRY_NUMBER, REQUEST_RETRY_WAIT_BACKOFF, async () => {
+            try {
+                const request = {
+                    method,
+                    params,
+                    id: (_nextId++),
+                    jsonrpc: '2.0'
+                };
+                const response = await fetchJson(this.connection, JSON.stringify(request));
+                if (response.error) {
+                    if (typeof response.error.data === 'object') {
+                        if (typeof response.error.data.error_message === 'string' && typeof response.error.data.error_type === 'string') {
+                            // if error data has error_message and error_type properties, we consider that node returned an error in the old format
+                            throw new TypedError(response.error.data.error_message, response.error.data.error_type);
+                        }
+                        
+                        throw parseRpcError(response.error.data);
+                    } else {
+                        const errorMessage = `[${response.error.code}] ${response.error.message}: ${response.error.data}`;
+                        // NOTE: All this hackery is happening because structured errors not implemented
+                        // TODO: Fix when https://github.com/nearprotocol/nearcore/issues/1839 gets resolved
+                        if (response.error.data === 'Timeout' || errorMessage.includes('Timeout error')
+                                || errorMessage.includes('query has timed out')) {
+                            throw new TypedError(errorMessage, 'TimeoutError');
+                        }
+
+                        throw new TypedError(errorMessage);
+                    }
                 }
-            } else {
-                const errorMessage = `[${response.error.code}] ${response.error.message}: ${response.error.data}`;
-                // NOTE: All this hackery is happening because structured errors not implemented
-                // TODO: Fix when https://github.com/nearprotocol/nearcore/issues/1839 gets resolved
-                if (response.error.data === 'Timeout' || errorMessage.includes('Timeout error')) {
-                    throw new TypedError('send_tx_commit has timed out.', 'TimeoutError');
-                } else {
-                    throw new TypedError(errorMessage);
+                return response.result;
+            } catch (error) {
+                // TODO: Somehow getting still:
+                // Error: send_tx_commit has timed out.
+                // TODO: Check if nearcore still sends this or some other messages for timeout
```

Comment:
@vgrichina have you checked this? probably it should be done before the merge

Reply:
Removed these TODOs as no longer relevant, `TimeoutError` thrown by us deeper in the call stack